packages/client/src/generators.test.ts

```diff
@@ -1,34 +1,113 @@
+    expect(verifier1).not.toEqual(verifier2);
+  });
 });
 
-test('generate codeChallenge with fixed length', async () => {
-  const verifier = generateCodeVerifier();
-  const challenge = await generateCodeChallenge(verifier);
-  expect(challenge.length).toEqual(Math.ceil(32 * 1.34));
-});
+describe('generateCodeChallenge', () => {
+  test('with fixed length', async () => {
+    const verifier = generateCodeVerifier();
+    const challenge = await generateCodeChallenge(verifier);
+    expect(challenge.length).toEqual(Math.ceil(32 * 1.34));
+  });
 
-test('generate codeChallenge with different codeVerifier', async () => {
-  const verifier1 = generateCodeVerifier();
-  const challenge1 = await generateCodeChallenge(verifier1);
-  const verifier2 = generateCodeVerifier();
-  const challenge2 = await generateCodeChallenge(verifier2);
-  expect(challenge1 === challenge2).toBeFalsy();
+  test('with different codeVerifier', async () => {
+    const verifier1 = generateCodeVerifier();
+    const challenge1 = await generateCodeChallenge(verifier1);
+    const verifier2 = generateCodeVerifier();
+    const challenge2 = await generateCodeChallenge(verifier2);
+    expect(challenge1).not.toEqual(challenge2);
+  });
+
+  test('with same codeVerifier', async () => {
+    const verifier = generateCodeVerifier();
+    const challenge1 = await generateCodeChallenge(verifier);
+    const challenge2 = await generateCodeChallenge(verifier);
+    expect(challenge1).toEqual(challenge2);
+  });
 });
 
-test('generate codeChallenge with same codeVerifier', async () => {
-  const verifier = generateCodeVerifier();
-  const challenge1 = await generateCodeChallenge(verifier);
-  const challenge2 = await generateCodeChallenge(verifier);
-  expect(challenge1 === challenge2).toBeTruthy();
+describe('generateScope', () => {
+  describe('input type : undefined', () => {
+    test('without parameters', () => {
+      expect(generateScope()).toEqual(DEFAULT_SCOPE_STRING);
+    });
+  });
+
+  describe('input type : string', () => {
+    test('empty', () => {
+      expect(generateScope('')).toEqual(DEFAULT_SCOPE_STRING);
+    });
+
+    test('with blank characters', () => {
+      expect(generateScope(' \t')).toEqual(DEFAULT_SCOPE_STRING);
+    });
+
+    test('with openid offline_access', () => {
+      const originalScope = DEFAULT_SCOPE_STRING;
+      expect(generateScope(originalScope)).toEqual(originalScope);
+    });
+
+    test('with openid (without default scope value offline_access)', () => {
+      expect(generateScope(OPENID)).toEqual(DEFAULT_SCOPE_STRING);
+    });
+
+    test('with openid name (without default scope value offline_access)', () => {
+      expect(generateScope(`${NAME} ${OPENID}`)).toEqual(`${DEFAULT_SCOPE_STRING} ${NAME}`);
+    });
+
+    test('with offline_access email (without default scope value openid)', () => {
+      expect(generateScope(`${EMAIL} ${OFFLINE_ACCESS}`)).toEqual(
+        `${DEFAULT_SCOPE_STRING} ${EMAIL}`
```

Comment:
Yes, the scope order matters.

The generateScope method always append or move the default scope values to the head of the scope string.

Reply:
I prefer the simpler implementation of the unit tests.

So I prefer to assert the result must equal the given constant string
rather than to call `contains` method (even though it's more robust when `generateScope ` implementation changes).